Sources/BSON/Integers+BSONValue.swift

```diff
@@ -1,6 +1,66 @@
 import NIO
 
 extension Int32: BSONValue {
+    /*
+    * Initializes an `Int32` from ExtendedJSON
+    *
+    * Parameters:
+    *   - `json`: a `JSON` representing the canonical or relaxed form of ExtendedJSON for an `Int32`.
+    *   - `keyPath`: an array of `String`s containing the enclosing JSON keys of the current json being passed in.
+    *              This is used for error messages.
+    *
+    * Returns:
+    *   - `nil` if the provided value is not an `Int32`.
+    *
+    * Throws:
+    *   - `DecodingError` if `json` is a partial match or is malformed.
+    */
+    internal init?(fromExtJSON json: JSON, keyPath: [String]? = nil) throws {
+        switch json {
+        case let .number(n):
+            // relaxed extended JSON
+            guard let int = Int32(exactly: n) else {
+                return nil
+            }
+            self = int
+        case let .object(obj):
+            // canonical extended JSON
+            guard let value = obj["$numberInt"] else {
+                return nil
+            }
+            guard let str = value.stringValue else {
+                throw DecodingError.dataCorrupted(
+                        DecodingError.Context(
+                                codingPath: [],
+                                debugDescription: "Expected the value: \"\(value)\" to be a string"
+                        )
+                )
+            }
+            guard obj.count == 1 else {
+                throw DecodingError.dataCorrupted(
+                    DecodingError.Context(
+                        codingPath: [],
+                        debugDescription: "Expected only \"$numberInt\" key, found extra keys: \(obj.keys)"
+                    )
+                )
+            }
+            guard let int = Int32(str) else {
+                let debugStart = keyPath != nil
+                        ? "\(keyPath!.joined(separator:".")): "  
+                        : ""
+                throw DecodingError.dataCorrupted(
+                    DecodingError.Context(
+                        codingPath: [],
+                        debugDescription: "\(debugStart): could not parse `Int32` from \"\(str)\"."
```

Comment:
We alway want this prefix appended to `DecodingError`s returned from this initializer, so that's why I was suggesting a helper function or extension of `DecodingError` to aid in that process. Then all three cases where we throw errors could just call through to that and automatically get the path appended properly.

Reply:
Similar to my comment above, we don't _always_ want to prefix the key path, but if it is present then we always want to. So the same suggestion applies but whatever helper would need to take in a `[String]?`.

Reply:
yep, I think I addressed this in the most recent commit :)